intlist.c

```diff
@@ -0,0 +1,269 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// intlist.c
+// Program on main branch.
+// ------------------------------------------------------------------------------------------------
+
+
+#include <stdio.h>
+#include <malloc.h>
+#include "intlist.h"
+
+
+struct node {
+   int value;
+   struct node* link;
+};
+
+//Defines head pointer and fixes it to null
+struct node* Create () {
+   struct node* head;
+   printf ("\nAn empty linked list is created \n");
+   head = NULL;
+   return head;
+}
+
+//Prints elements of the list starting from the last element that was added to the list
+void PrintFunc (struct node* s) {
+   struct node* temp = s;
+   if (temp != NULL) {
+      printf ("\nThe list elements are : \n");
+      while (temp != NULL) {
+         printf ("%d -> ", temp->value);
+         temp = temp->link;
+      }
+      printf ("\n");
+   } else { printf ("\nNo elements in the list\n"); }
+}
+
+//Adds element at the end of the list
+void Add (struct node** head, int val) {
+   //Adds element to an empty list
+   if (head == NULL) {
+      struct node* temp = malloc (sizeof (struct node));
+      temp->value = val;
+      temp->link = NULL;
+      *head = temp;
+   }
+   struct node* result = malloc (sizeof (struct node));
+   result->value = val;
+   result->link = *head; //current struct head pointer is linked to the new element
+   *head = result;// struct head pointer now points to the added element 
+   struct node* store = *head;
+   printf ("\n%d is added to the list\n", (*head)->value);
+   PrintFunc (store);
+   return;
+}
+
+//Deletes the entire list
+void Delete (struct node** head) {
+   struct node* store = *head;
+   if (store == NULL) {
+      printf ("\nThere are no elements to delete \n");
+      return;
+   } else {
+      struct node* temp;
+      struct node* temp2;
+      temp = store;
+      while (temp != NULL) {
+         printf ("\n%d is deleted from the list\n", temp->value);
+         temp2 = temp->link;
+         free (temp);
+         temp = NULL;
+         temp = temp2;
+      }
+      printf ("\nList deleted... \n");
+      *head = NULL;
+      return;
+   }
+}
+
+//Prints the number of elements in the list
+void Count (struct node** head) {
+   struct node* store = *head;

```

Comment:
What is the need for the double indirection i.e. **?

Reply:
Since the operation might require updating the head pointer and the program might have more than one linked list I used double indirection. As per TA's suggestion I have made some changes to the same to the best of my knowledge. Thank you.